[Detection Engine] Fix infinite-looping bug related to bootstrapping lists resources

rylnd opened the pull request:
# Summary

Addresses https://github.com/elastic/kibana/issues/229018.

## Problem Statement
When the users visits the alerting page, we check to see whether the lists indices exist, and if not, and they have the appropriate privileges, we try to create it on behalf of the user.

However, if the user has enough privileges to _ask_ about the lists indices status, but has insufficient privileges to _create_ the lists indices, the frontend will enter an infinite loop. With sufficient privileges, this loop is exited when the frontend creates the lists indices.

## Explanation
The looping is due to a) the fact that the "read list index" API responds with a 404 if either of the indices is not present, and b) how 404s are handled in the frontend. By default, a 404 will cause the http client to throw an error. We use useQuery to cache this query, but because the request throws an error the request is never cached, and thus on the next hook render the cycle renews.

## Solution
By manually catching and parsing the 404 API response, we can both a) ensure proper caching and b) present more information to the consumer of that API. For backwards compatibility reasons, we cannot update the semantics of the API itself.

Practically, since the response to a "list index not found" response is to call the (generic) "create list index" API, this additional information is not strictly necessary, the additional information could e.g. be used to provide a more helpful error message on the client in the future.

## Further Improvements
In these cases where the lists indices do not exist, and the user does not have permission to create them, we do not currently inform the user of this fact. We could improve the user experience by providing a more helpful error message in this scenario.


# How to review
1. Create a user/role in kibana with _almost_ all of the [necessary privileges to create the list indices](https://www.elastic.co/docs/solutions/security/detect-and-alert/detections-requirements#security-detections-requirements-custom-role-privileges), save for the "manage" cluster privilege (which is easily overlooked, due to its being a single word in a column on that table): 

    ```json
    almost-lists-manager:
      cluster: []
      indices:
        - names:
            - ".alerts-security.alerts-default"
            - ".internal.alerts-security.alerts-default-*"
            - ".siem-signals-default"
            - ".lists-default"
            - ".items-default"
          privileges: ["manage", "read", "write", "view_index_metadata"]
          field_security:
            grant: ["*"]
            except: []
          allow_restricted_indices: false
        - names:
            - ".preview.alerts-security.alerts-default"
            - ".internal.preview.alerts-security.alerts-default-*"
          privileges: ["read"]
          field_security:
            grant: ["*"]
            except: []
          allow_restricted_indices: false
      applications:
        - application: "kibana-.kibana"
          privileges: ["feature_indexPatterns.all", "feature_siemV4.all", "feature_actions.all", "feature_savedObjectsManagement.all"]
          resources: ["space:default"]
      run_as: []
      metadata: {}
      transient_metadata:
        enabled: true
      description: ""
    ```

1. Navigate to the alerts page, and confirm that the page loads successfully, and does not loop.
1. (optional) Confirm that navigating to the alerts page on `main` reproduces the infinite loop.
1. (optional) Confirm that adding the "manage" cluster privilege results in the lists indices being created successfully.

### Checklist


- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.




<!--ONMERGE {"backportTargets":["8.19","9.2"]} ONMERGE-->

rylnd committed:
Fix infinite-looping bug related to bootstrapping lists resources

Addresses https://github.com/elastic/kibana/issues/229018.

When the users visits the alerting page, we check to see whether the
lists indices exist, and if not, and they have the appropriate
privileges, we try to create it on behalf of the user.

However, if the user has enough privileges to _ask_ about the lists
indices status, but has insufficient privileges to _create_ the lists
indices, the frontend will enter an infinite loop. With sufficient
privileges, this loop is exited when the frontend creates the lists
indices.

The looping is due to a) the fact that the "read list index" API
responds with a 404 if either of the indices is not present, and b) how
404s are handled in the frontend. By default, a 404 will cause the http
client to throw an error. We use useQuery to cache this query, but
because the request throws an error the request is never cached, and
thus on the next hook render the cycle renews.

By manually catching and parsing the 404 API response, we can both a)
ensure proper caching and b) present more information to the consumer of
that API. For backwards compatibility reasons, we cannot update the
semantics of the API itself.

Practically, since the response to a "list index not found" response is
to call the (generic) "create list index" API, this additional
information is not strictly necessary, the additional information could
e.g. be used to provide a more helpful error message on the client in
the future.

rylnd committed:
Cleanup: remove dead code

The previous commit captures this error nearer to the call site;
handling the error at the caller is no longer necessary.

kibanamachine committed:
Changes from node scripts/lint_packages --fix

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

rylnd committed:
Include TSX files in compliation of list-hooks package

We added a test file that exercises one of these hooks; it required
creating a component wrapper to mock the use of `useQuery` in the hook.

rylnd committed:
Merge branch 'main' into fix-lists-loop-bug

rylnd committed:
Merge branch 'main' into fix-lists-loop-bug

rylnd committed:
Merge branch 'main' into fix-lists-loop-bug

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/356496)
* Commit: 1b63ee8fe1b00406eec9712e11822817ed9ee8b8

### Failed CI Steps
* [Build Kibana Distribution](https://buildkite.com/elastic/kibana-pull-request/builds/356496#019a3241-4da0-498a-9970-9bef98dfa329)

### History
* :broken_heart: [Build #356492](https://buildkite.com/elastic/kibana-pull-request/builds/356492) failed e627e9094b2321fce2b14cf4216befe88ba3b9c3
* :broken_heart: [Build #356475](https://buildkite.com/elastic/kibana-pull-request/builds/356475) failed e627e9094b2321fce2b14cf4216befe88ba3b9c3
* :broken_heart: [Build #356432](https://buildkite.com/elastic/kibana-pull-request/builds/356432) failed 86b5a97ec95fb037a5168508ea03d4c1488cdec8
* :broken_heart: [Build #356311](https://buildkite.com/elastic/kibana-pull-request/builds/356311) failed aa7f244c921d21a13108189b0cf00209334e2803
* :broken_heart: [Build #356023](https://buildkite.com/elastic/kibana-pull-request/builds/356023) failed 556a4efbd5ebe47b0c3f8f6e633fa86ff683bfe1
* :broken_heart: [Build #356021](https://buildkite.com/elastic/kibana-pull-request/builds/356021) failed 453ae1feaa33f95f77b16a79b9fe7214856a0e09

cc @rylnd

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356496","number":356496,"commit":"1b63ee8fe1b00406eec9712e11822817ed9ee8b8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356492","number":356492,"commit":"e627e9094b2321fce2b14cf4216befe88ba3b9c3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356475","number":356475,"commit":"e627e9094b2321fce2b14cf4216befe88ba3b9c3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356432","number":356432,"commit":"86b5a97ec95fb037a5168508ea03d4c1488cdec8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356311","number":356311,"commit":"aa7f244c921d21a13108189b0cf00209334e2803"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356023","number":356023,"commit":"556a4efbd5ebe47b0c3f8f6e633fa86ff683bfe1"}],"number":356496}
buildkite-pr-comment-->

rylnd committed:
Merge branch 'main' into fix-lists-loop-bug

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/356520)
* Commit: 73e808407b8d5144a4d4e21f3388501d5c31fd6e

### Failed CI Steps
* [Investigations - Security Solution Cypress Tests #4](https://buildkite.com/elastic/kibana-pull-request/builds/356520#019a3344-5d4b-41b6-baa7-a8d27532317e)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/356520#019a3344-5d4b-41b6-baa7-a8d27532317e) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/356520/jobs/019a3344-5d4b-41b6-baa7-a8d27532317e/artifacts/019a337e-7b8d-44f0-b8b4-ae5f02ea2f21) Investigations - Security Solution Cypress Tests #<span></span>4 / Alert Table API calls "before each" hook for "should call `api/lists/index` only once" "before each" hook for "should call `api/lists/index` only once"

### History
* :broken_heart: [Build #356496](https://buildkite.com/elastic/kibana-pull-request/builds/356496) failed 1b63ee8fe1b00406eec9712e11822817ed9ee8b8
* :broken_heart: [Build #356492](https://buildkite.com/elastic/kibana-pull-request/builds/356492) failed e627e9094b2321fce2b14cf4216befe88ba3b9c3
* :broken_heart: [Build #356475](https://buildkite.com/elastic/kibana-pull-request/builds/356475) failed e627e9094b2321fce2b14cf4216befe88ba3b9c3
* :broken_heart: [Build #356432](https://buildkite.com/elastic/kibana-pull-request/builds/356432) failed 86b5a97ec95fb037a5168508ea03d4c1488cdec8
* :broken_heart: [Build #356311](https://buildkite.com/elastic/kibana-pull-request/builds/356311) failed aa7f244c921d21a13108189b0cf00209334e2803
* :broken_heart: [Build #356023](https://buildkite.com/elastic/kibana-pull-request/builds/356023) failed 556a4efbd5ebe47b0c3f8f6e633fa86ff683bfe1

cc @rylnd

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356520","number":356520,"commit":"73e808407b8d5144a4d4e21f3388501d5c31fd6e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356496","number":356496,"commit":"1b63ee8fe1b00406eec9712e11822817ed9ee8b8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356492","number":356492,"commit":"e627e9094b2321fce2b14cf4216befe88ba3b9c3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356475","number":356475,"commit":"e627e9094b2321fce2b14cf4216befe88ba3b9c3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356432","number":356432,"commit":"86b5a97ec95fb037a5168508ea03d4c1488cdec8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356311","number":356311,"commit":"aa7f244c921d21a13108189b0cf00209334e2803"}],"number":356520}
buildkite-pr-comment-->

vitaliidm commented on x-pack/solutions/security/packages/kbn-securitysolution-list-hooks/src/use_read_list_index/index.ts:
```diff
@@ -48,3 +60,41 @@ export const useReadListIndex = ({
+export const parseReadIndexResultFrom404Error = (
+  error: SecurityAppError
+): Awaited<ReturnType<typeof readListIndexWithOptionalSignal>> => {
+  if (error.body.message.includes(' and ')) {
```
this seems like very generic error and without any context.

If we expect certain wording, maybe it makes sense to include it?
Or use some kind of error code

vitaliidm commented on x-pack/solutions/security/packages/kbn-securitysolution-list-hooks/src/use_read_list_index/index.ts:
```diff
@@ -48,3 +60,41 @@ export const useReadListIndex = ({
+      list_index: false,
+      list_item_index: false,
+    };
+  } else if (error.body.message.includes('item')) {
```
similar here

rylnd committed:
Merge branch 'main' into fix-lists-loop-bug

rylnd committed:
More robust and simpler error parsing logic

Feedback from review.

rylnd committed:
Only parse 404 message if it is of the expected form

Without this guard, we previously would have parsed any 404 response
coming from our APIs, which with the previous logic would result in
`true/true` for most errors.

Instead, we now move those 404s to the "error" case. Added tests to
capture this behavior.

rylnd committed:
Silence error logs in tests

By default the QueryClient will log errors.

rylnd commented on x-pack/solutions/security/packages/kbn-securitysolution-list-hooks/src/use_read_list_index/index.ts:
```diff
@@ -48,3 +60,41 @@ export const useReadListIndex = ({
+export const parseReadIndexResultFrom404Error = (
+  error: SecurityAppError
+): Awaited<ReturnType<typeof readListIndexWithOptionalSignal>> => {
+  if (error.body.message.includes(' and ')) {
```
[ebe2e50](https://github.com/elastic/kibana/pull/241052/commits/ebe2e50764086a2e468417fa2b069ee3869012df)

rylnd committed:
Remove duplicated header

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/356999)
* Commit: a8101724fb43f41a73d02c5c2883ace8fef7b45a

### Failed CI Steps
* [FTR Configs #66](https://buildkite.com/elastic/kibana-pull-request/builds/356999#019a371e-2f6c-4ae1-adeb-393e8ff796ba)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/356999#019a371e-2f6c-4ae1-adeb-393e8ff796ba) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/356999/jobs/019a371e-2f6c-4ae1-adeb-393e8ff796ba/artifacts/019a3747-ba43-43e5-8483-2fd051fa89b2) FTR Configs #<span></span>66 / Dataset Quality Dataset quality handles user privileges User can read logs-* User cannot monitor any data stream "before all" hook for "Estimated data are not available due to underprivileged user"

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/1b3e9e172e7437a8505de80826ba73e1c6df97d7) | [after](https://github.com/elastic/kibana/commit/a8101724fb43f41a73d02c5c2883ace8fef7b45a) | diff |
| --- | --- | --- | --- |
| `lists` | 338 | 345 | +7 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/1b3e9e172e7437a8505de80826ba73e1c6df97d7) | [after](https://github.com/elastic/kibana/commit/a8101724fb43f41a73d02c5c2883ace8fef7b45a) | diff |
| --- | --- | --- | --- |
| `@kbn/securitysolution-list-hooks` | 77 | 78 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/1b3e9e172e7437a8505de80826ba73e1c6df97d7) | [after](https://github.com/elastic/kibana/commit/a8101724fb43f41a73d02c5c2883ace8fef7b45a) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 11.0MB | 11.0MB | +186.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/1b3e9e172e7437a8505de80826ba73e1c6df97d7) | [after](https://github.com/elastic/kibana/commit/a8101724fb43f41a73d02c5c2883ace8fef7b45a) | diff |
| --- | --- | --- | --- |
| `@kbn/securitysolution-list-hooks` | 88 | 92 | +4 |


</details>


### History
* :broken_heart: [Build #356520](https://buildkite.com/elastic/kibana-pull-request/builds/356520) failed 73e808407b8d5144a4d4e21f3388501d5c31fd6e
* :broken_heart: [Build #356496](https://buildkite.com/elastic/kibana-pull-request/builds/356496) failed 1b63ee8fe1b00406eec9712e11822817ed9ee8b8
* :broken_heart: [Build #356492](https://buildkite.com/elastic/kibana-pull-request/builds/356492) failed e627e9094b2321fce2b14cf4216befe88ba3b9c3
* :broken_heart: [Build #356475](https://buildkite.com/elastic/kibana-pull-request/builds/356475) failed e627e9094b2321fce2b14cf4216befe88ba3b9c3
* :broken_heart: [Build #356432](https://buildkite.com/elastic/kibana-pull-request/builds/356432) failed 86b5a97ec95fb037a5168508ea03d4c1488cdec8

cc @rylnd

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356999","number":356999,"commit":"a8101724fb43f41a73d02c5c2883ace8fef7b45a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356520","number":356520,"commit":"73e808407b8d5144a4d4e21f3388501d5c31fd6e"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356496","number":356496,"commit":"1b63ee8fe1b00406eec9712e11822817ed9ee8b8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356492","number":356492,"commit":"e627e9094b2321fce2b14cf4216befe88ba3b9c3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356475","number":356475,"commit":"e627e9094b2321fce2b14cf4216befe88ba3b9c3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356432","number":356432,"commit":"86b5a97ec95fb037a5168508ea03d4c1488cdec8"}],"number":356999}
buildkite-pr-comment-->

vitaliidm reviewed (approved):
thanks for addressing the comments

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.19|Backport failed because of merge conflicts|
|❌|9.2|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 241052
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

rylnd commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.2|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/241496">](https://github.com/elastic/kibana/pull/241496)|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/241497">](https://github.com/elastic/kibana/pull/241497)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.1.0 -->
